File: src/utils/initialState/ship.ts
import type { ActionDamage, Actions } from '@nanosh/types/actions'
import type { CharacterNames } from '@nanosh/types/character'
import {
  CANNON_DEFAULT_MAX_DAMAGE,
  CANNON_DEFAULT_MIN_DAMAGE,
} from '@nanosh/types/generic'
import type { ModifierTracker, ModifiersShip } from '@nanosh/types/modifiers'
import type { ProjectNames, ProjectProgress } from '@nanosh/types/projects'
import { type Ship, type ShipCargo, type ShipTypes } from '../../types/ship'
import { GetInitialExpoCrafts, GetInitialFighterCrafts } from './aircrafts'
import { GetInitialShipRooms } from './rooms'

export interface GetInitialShipParams {
  type: ShipTypes
}
export const GetInitialShip = (
  { type }: GetInitialShipParams = { type: 'griffin' },
): Ship => {
  const ship: Ship = {
    type,
    health: 0,
    maxHealth: 0,
    cargo: new Set<ShipCargo>(),
    maxCargoSize: 0,
    eCells: 0,
    supplies: 0,
    rations: 0,
    civitates: 0,
    praetorians: 0,
    rooms: GetInitialShipRooms({ shipType: type }),
    fighterCrafts: GetInitialFighterCrafts({ shipType: type }),
    expoCrafts: GetInitialExpoCrafts({ shipType: type }),
    expo: {
      members: new Set<CharacterNames>(),
    },
    modifiers: new Map<ModifiersShip, ModifierTracker>(),
    projects: {
      queued: new Map<ProjectNames, ProjectProgress>(),
      done: new Set<ProjectNames>(),
      pool: new Set<ProjectNames>([
        'File G11 - Apex Bio Enhancement',
        'File 128 - Finesse Protocol',
        'File 129 - Equilibrium Drive',
        'File 252 - HyperHeal Ampoule',
        'File 253 - Lifesaver Initiative',
        'File 254 - Operational Surge Paradigm',
        'File 311 - Provisioning Overhaul',
        'File E120 - Solo Comfort Initiative',
        'File 100 - Support Blueprints Recovery A',
        'File 101 - Support Blueprints Recovery B',
        'File 112 - EisenSchlag Modul',
        'File 113 - Biogenic Aim Assist',
        'File 456 - Kabuto Boost',
        'File 010 - M22 "Buzzard" Precision Upgrade',
        'File 011 - M22 "Buzzard" Strike Optimization',
        'File 012 - M22 "Buzzard" Defensive Retrofit',
        'File 055 - Hoppers Space Optimization',
        'File 711 - Praetorians Suit Force Distribution',
        "File 712 - Ysara's Snare",
        'File NAP - Nanosh Assimilation Protocol',
      ]),
    },
    damage: new Map<Actions, ActionDamage>([
      [
        'action.bridge.command.cannon',
        { min: CANNON_DEFAULT_MIN_DAMAGE, max: CANNON_DEFAULT_MAX_DAMAGE },
      ],
    ]),
  }

  switch (type) {
    case 'griffin':
      ship.health = 200
      ship.maxHealth = 300
      ship.maxCargoSize = 800
      ship.eCells = 120
      ship.supplies = 240
      ship.rations = 100
      ship.civitates = 10
      ship.praetorians = 5
      break
  }

  return ship
}

File: src/nakama/gamelogic/utils/initial/ship.go
package initial

import "nanosh/nakama-modules/gamelogic/types"

// GetInitialShipParams defines parameters for getting initial ship
type GetInitialShipParams struct {
	Type types.ShipType `json:"type"`
}

// GetInitialShip creates and returns initial ship state for a given ship type
// Defaults to griffin if no ship type provided
func GetInitialShip(params *GetInitialShipParams) *types.Ship {
	// Default to griffin if no params provided
	if params == nil {
		params = &GetInitialShipParams{Type: types.ShipGriffin}
	}

	// Initialize ship with zero values
	ship := &types.Ship{
		Type:          params.Type,
		Health:        0,
		MaxHealth:     0,
		Shield:        0,
		MaxShield:     0,
		Cargo:         make(map[string]*types.ShipCargo), // Empty cargo map
		MaxCargoSize:  0,
		ECells:        0,
		Supplies:      0,
		Rations:       0,
		Civitates:     0,
		Praetorians:   0,
		Rooms:         GetInitialShipRooms(&GetInitialShipRoomsParams{ShipType: params.Type}),
		FighterCrafts: GetInitialFighterCrafts(&GetInitialAircraftsParams{ShipType: params.Type}),
		ExpoCrafts:    GetInitialExpoCrafts(&GetInitialAircraftsParams{ShipType: params.Type}),
		Expo: types.Expo{
			Members: make(map[types.CharacterName]bool), // Empty members set
		},
		Modifiers: make(map[types.ModifierShip]*types.ModifierTracker), // Empty modifiers
		Projects: types.ProjectPool{
			Queued: make(map[types.ProjectName]*types.ProjectProgress), // Empty queued projects
			Done:   make(map[types.ProjectName]bool),                   // Empty done projects set
			Pool:   make(map[types.ProjectName]bool),                   // Will be filled below
		},
		Damage: make(map[types.Action]*types.MinMaxStruct), // Will be filled below
	}

	// Initialize project pool with all available projects
	projectPool := []types.ProjectName{
		types.ProjectFileG11ApexBioEnhancement,
		types.ProjectFile128FinesseProtocol,
		types.ProjectFile129EquilibriumDrive,
		types.ProjectFile252HyperHealAmpoule,
		types.ProjectFile253LifesaverInitiative,
		types.ProjectFile254OperationalSurgeParadigm,
		types.ProjectFile311ProvisioningOverhaul,
		types.ProjectFileE120SoloComfortInitiative,
		types.ProjectFile100SupportBlueprintsRecoveryA,
		types.ProjectFile101SupportBlueprintsRecoveryB,
		types.ProjectFile112EisenSchlagModul,
		types.ProjectFile113BiogenicAimAssist,
		types.ProjectFile456KabutoBoost,
		types.ProjectFile010BuzzardPrecisionUpgrade,
		types.ProjectFile011BuzzardStrikeOptimization,
		types.ProjectFile012BuzzardDefensiveRetrofit,
		types.ProjectFile055HoppersSpaceOptimization,
		types.ProjectFile711PraetoriansSuitForceDistribution,
		types.ProjectFile712YsarasSnare,
		types.ProjectFileNAPNanoshAssimilationProtocol,
	}

	// Add all projects to pool
	for _, project := range projectPool {
		ship.Projects.Pool[project] = true
	}

	// Initialize cannon damage
	ship.Damage[types.ActionBridgeCommandCannon] = &types.MinMaxStruct{
		Min: types.CANNON_DEFAULT_MIN_DAMAGE,
		Max: types.CANNON_DEFAULT_MAX_DAMAGE,
	}

	// Set ship-specific values
	switch params.Type {
	case types.ShipGriffin:
		ship.Health = 200
		ship.MaxHealth = 300
		ship.MaxShield = 50
		ship.MaxCargoSize = 800
		ship.ECells = 120
		ship.Supplies = 240
		ship.Rations = 100
		ship.Civitates = 10
		ship.Praetorians = 5
	}

	return ship
}
